### PR TITLE
Hole Punching Protocol

### PR DESCRIPTION
**Pull request by vyzo**
For https://github.com/libp2p/go-libp2p/issues/1039.

Implements the Hole Punching co-ordination protocol and upgrades dependencies to solve multi-stream simultaneous connect.

### TODO
- [ ] Deps.
- [ ] Allow the Hole Punching Stream to open on a transient connection.
- [ ] Unit Tests (once we are happy with the general direction we are taking here).

### CODE REVIEW

**Comment by aarshkshah1992** [comment]
Ping @marten-seemann @Stebalien for review.

**Comment by aarshkshah1992** [comment]
ping @vyzo for review .

File: p2p/protocol/holepunch/coordination.go [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+		return
+	}
+	defer s.Close()
+
+	// Hole punch now by forcing a connect
+	pi := peer.AddrInfo{
+		ID:    rp,
+		Addrs: obsDial,
+	}
+	hs.holePunchConnectWithBackoff(pi)
```
Comment (aarshkshah1992) [inline p2p/protocol/holepunch/coordination.go:192]: @Stebalien @vyzo @marten-seemann 

I am a bit concerned about the user experience here. Say `user1` calls `host.Connect`with `user2` to open a `Bitswap` stream later on. Assume both `user1` and `user2` are NATT'd.

Now, the `Host.Connect` call will return as soon as `user1` has a Relayed connection to `user2`. The Hole Punching happens in the background and the user needs to watch out for a direct connection by listening on the `Connected notification` emitted by the Swarm.

Is it worth introducing a `DirectConnect` or similar API on the `Host` that blocks till we have a direct/hole-punched connection with the remote peer and returns a failure if the hole punch fails ?

File: p2p/protocol/holepunch/coordination.go [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+		hs.refCount.Wait()
+	})
+
+	return nil
+}
+
+// attempts to make a direct connection with the remote peer of `relayConn` by co-ordinating a hole punch over
+// the given relay connection `relayConn`.
+func (hs *HolePunchService) holePunch(relayConn network.Conn) {
+	rp := relayConn.RemotePeer()
```
Comment (aarshkshah1992) [inline p2p/protocol/holepunch/coordination.go:69]: Same concern as https://github.com/libp2p/go-libp2p/pull/711/files#r561681671.

**Comment by requilence** [comment]
@aarshkshah1992 Hello! I'm looking forward to this issue being merged, as we are facing a lot of troubles with multiplexing with TCP transport.

BTW, I've just run the bidirectional connections test from this issue: https://github.com/libp2p/go-libp2p-noise/issues/70, and it fails: the `h1.Connect(...)` returns nil err but the next-line `h1.Network().Peers()` returns 0 peers.
The same happens when running with NoSecurity option.

**Comment by aarshkshah1992** [comment]
@requilence That definitely sounds like a bug. I'll take a look !

**Comment by aarshkshah1992** [comment]
@requilence There was a race in that test wherein a peer could check for peers it is connected to after that connection has been closed by the remote peer depending on when it hit that code path.

I've fixed the test and you can follow this issue at https://github.com/libp2p/go-libp2p/pull/1041.  Once this PR is merge, we can release simopen.

**Comment by requilence** [comment]
@aarshkshah1992 can I clarify? Do both of them share the same connection? For each node, it seems that the connection has been initiated on their side and can be safely closed when the operation on their side is finished. 

Here is an example
- node A opens a connection to node B and starts some long transfer via stream
- node B opens a connection to node B and downloads a file via bitswap
- node B finish blocks exchange and closes the connection
- will it lead to stream initiated by node A will be reset?

File: p2p/protocol/holepunch/pb/holepunch.proto [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  }
+
+  optional Type type=1;
+  // For hole punching, we'll send some additional observed addresses to the remote peer
+  // that could have been filtered by the Host address factory (for example: AutoRelay removes all public addresses if peer has private reachability).
+  // This is a hack!
+  // We plan to have a better address discovery and advertisement mechanism in the future.
+  // See https://github.com/libp2p/go-libp2p-autonat/pull/98
+  repeated bytes ObsAddrs = 2;
+}
```
Comment (vyzo) [inline p2p/protocol/holepunch/pb/holepunch.proto:18]: newline....

**Review by vyzo — COMMENTED** [review]
this is looking pretty good, a few comments.

Note: I can't green tick this one, as I originally opened the pr.

**Comment by vyzo** [comment]
Note that this is getting quite messy in terms of commit history, we might have to squash it.

File: p2p/protocol/holepunch/coordination.go [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+	"github.com/libp2p/go-msgio/protoio"
+	ma "github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr/net"
+)
+
+// TODO Should we have options for these ?
+const (
+	protocol         = "/libp2p/holepunch/1.0.0"
+	maxMsgSize       = 4 * 1024 // 4K
+	holePunchTimeout = 2 * time.Minute
```
[the remark this answers is not in the document]
Comment (vyzo) [reply p2p/protocol/holepunch/coordination.go:24]: It's 1min currently.

**Comment by vyzo** [comment]
We will want to integrate the circuit v2 pr here and change the libp2p constructor to use the new beast instead of the legacy one.

File: p2p/protocol/holepunch/coordination.go [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+			if err := hs.host.Connect(dialCtx, peer.AddrInfo{ID: rp}); err == nil {
+				log.Debugf("direct connection to peer %s successful, no need for a hole punch", rp.Pretty())
+				return
+			}
+			break
+		}
+	}
+
+	// hole punch
+	s, err := hs.host.NewStream(hs.ctx, rp, protocol)
```
[the remark this answers is not in the document]
Comment (aarshkshah1992) [reply p2p/protocol/holepunch/coordination.go:88]: done.

File: p2p/protocol/holepunch/coordination.go [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+	rd := protoio.NewDelimitedReader(s, maxMsgSize)
+	msg.Reset()
+	if err := rd.ReadMsg(msg); err != nil {
+		s.Reset()
+		log.Errorf("failed to read connect message from remote peer, err: %s", err)
+		return
+	}
+	if msg.GetType() != holepunch_pb.HolePunch_CONNECT {
+		s.Reset()
+		return
```
[the remark this answers is not in the document]
Comment (aarshkshah1992) [reply p2p/protocol/holepunch/coordination.go:117]: done.

File: p2p/protocol/holepunch/pb/holepunch.proto (hunk 3 above)
Comment (aarshkshah1992) [reply p2p/protocol/holepunch/pb/holepunch.proto:18]: done.

**Comment by aarshkshah1992** [comment]
Closed in favour of https://github.com/libp2p/go-libp2p/pull/1057.

**Pull request closed without merging**